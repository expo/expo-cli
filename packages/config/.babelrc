{
  "presets": [
    [
      "env",
      {
        "targets": {
          "node": "6.0.0"
        }
      }
    ]
  ],
  "plugins": [
    "transform-runtime",
    ["transform-object-rest-spread", { "useBuiltins": true }],
    "transform-flow-strip-types"
  ]
}
