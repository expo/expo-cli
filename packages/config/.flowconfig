[ignore]
<PROJECT_ROOT>/build/.*
.*/eslint-config-universe/.*

[options]
emoji=true

esproposal.decorators=ignore
esproposal.class_static_fields=enable
esproposal.class_instance_fields=enable
esproposal.export_star_as=enable

module.ignore_non_literal_requires=true

unsafe.enable_getters_and_setters=true

suppress_comment=.*\\$FlowFixMe
suppress_comment=.*\\$FlowIssue
suppress_comment=\\(.\\|\n\\)*\\$FlowNewLine
suppress_comment=\\(.\\|\n\\)*\\$ExpectError

