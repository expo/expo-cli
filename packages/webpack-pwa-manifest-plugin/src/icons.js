import fs from 'fs-extra';
import Jimp from 'jimp';
import mime from 'mime';
import { joinURI } from './helpers/uri';
import generateFingerprint from './helpers/fingerprint';
import IconError from './errors/IconError';
import { fromStartupImage } from './validators/Apple';

const supportedMimeTypes = [Jimp.MIME_PNG, Jimp.MIME_JPEG, Jimp.MIME_BMP];

const ASPECT_FILL = 'cover';
const ASPECT_FIT = 'contain';

export async function createBaseImageAsync(width, height, color) {
  return new Promise(
    (resolve, reject) =>
      new Jimp(width, height, color, (err, image) => {
        if (err) {
          reject(err);
          return;
        }
        resolve(image);
      })
  );
}

async function compositeImagesAsync(image, ...images) {
  for (const imageProps of images) {
    const childImage = await Jimp.read(imageProps);
    image.composite(childImage, 0, 0);
  }
  return image;
}

function parseArray(i) {
  if (i == null) return [];
  return i && !Array.isArray(i) ? [i] : i;
}

function sanitizeIcon(iconSnippet) {
  if (!iconSnippet.src) {
    throw new IconError('Unknown icon source.');
  }
  const sizes = parseArray(iconSnippet.size || iconSnippet.sizes);
  if (!sizes) {
    throw new IconError('Unknown icon sizes.');
  }
  return {
    src: iconSnippet.src,
    resizeMode: iconSnippet.resizeMode,
    sizes,
    media: iconSnippet.media,
    destination: iconSnippet.destination,
    ios: iconSnippet.ios || false,
    color: iconSnippet.color,
  };
}

function processIcon(width, height, icon, buffer, mimeType, publicPath, shouldFingerprint) {
  const dimensions = `${width}x${height}`;
  const fileName = shouldFingerprint
    ? `icon_${dimensions}.${generateFingerprint(buffer)}.${mime.getExtension(mimeType)}`
    : `icon_${dimensions}.${mime.getExtension(mimeType)}`;
  const iconOutputDir = icon.destination ? joinURI(icon.destination, fileName) : fileName;
  const iconPublicUrl = joinURI(publicPath, iconOutputDir);
  return {
    manifestIcon: {
      src: iconPublicUrl,
      sizes: dimensions,
      type: mimeType,
    },
    webpackAsset: {
      output: iconOutputDir,
      url: iconPublicUrl,
      source: buffer,
      size: buffer.length,
      ios: icon.ios
        ? { valid: icon.ios, media: icon.media, size: dimensions, href: iconPublicUrl }
        : false,
      resizeMode: icon.resizeMode,
      color: icon.color,
    },
  };
}

function parseSize(size) {
  let width;
  let height;
  if (Array.isArray(size) && size.length) {
    // [0, 0] || [0]
    width = size[0];
    height = size.length > 1 ? size[1] : size[0];
  } else if (typeof size === 'number') {
    // 0
    width = size;
    height = size;
  } else if (typeof size === 'string') {
    // '0x0'
    const dimensions = size.split('x');
    width = dimensions[0];
    height = dimensions[1];
  }
  return { width, height };
}

async function getBufferWithMimeAsync({ src, resizeMode, color }, mimeType, { width, height }) {
  if (!supportedMimeTypes.includes(mimeType)) {
    try {
      return fs.readFileSync(src);
    } catch (err) {
      throw new IconError(`It was not possible to read '${src}'.`);
    }
  } else {
    return await resize(src, mimeType, width, height, resizeMode, color);
  }
}

async function processImage(size, icon, fingerprint, publicPath) {
  const { width, height } = parseSize(size);
  if (width <= 0 || height <= 0) {
    return;
  }
  const mimeType = mime.getType(icon.src);
  const _buffer = await getBufferWithMimeAsync(icon, mimeType, { width, height });
  return processIcon(width, height, icon, _buffer, mimeType, publicPath, fingerprint);
}

async function resize(img, mimeType, width, height, resizeMode = 'contain', color) {
  try {
    const initialImage = await Jimp.read(img);
    const center = Jimp.VERTICAL_ALIGN_MIDDLE | Jimp.HORIZONTAL_ALIGN_CENTER;
    if (resizeMode === ASPECT_FILL) {
      return await initialImage
        .cover(width, height, center)
        .quality(100)
        .getBufferAsync(mimeType);
    } else if (resizeMode === ASPECT_FIT) {
      const resizedImage = await initialImage.contain(width, height, center).quality(100);
      if (!color) {
        return resizedImage.getBufferAsync(mimeType);
      }

      const splashScreen = await createBaseImageAsync(width, height, color);
      const combinedImage = await compositeImagesAsync(splashScreen, resizedImage);
      return combinedImage.getBufferAsync(mimeType);
    } else {
      throw new IconError(
        `Unsupported resize mode: ${resizeMode}. Please choose either 'cover', or 'contain'`
      );
    }
  } catch ({ message }) {
    throw new IconError(`It was not possible to generate splash screen '${img}'. ${message}`);
  }
}

export function retrieveIcons(manifest) {
  // Remove these items so they aren't written to disk.
  const { startupImages, apple, icon, icons, ...config } = manifest;
  const parsedStartupImages = parseArray(startupImages);

  let parsedIcons = parseArray(icons);

  if (parsedStartupImages.length) {
    // TODO: Bacon: use all of the startup images
    const startupImage = parsedStartupImages[0];
    parsedIcons = [...parsedIcons, ...fromStartupImage(startupImage)];
  }

  const response = parsedIcons.map(icon => sanitizeIcon(icon));
  return [response, config];
}

export async function parseIcons(inputIcons, fingerprint, publicPath) {
  if (!inputIcons.length) {
    return {};
  }

  let icons = [];
  let assets = [];

  let promises = [];
  for (const icon of inputIcons) {
    const { sizes } = icon;
    promises = [
      ...promises,
      ...sizes.map(async size => {
        const { manifestIcon, webpackAsset } = await processImage(
          size,
          icon,
          fingerprint,
          publicPath
        );
        icons.push(manifestIcon);
        assets.push(webpackAsset);
      }),
    ];
  }
  await Promise.all(promises);

  return {
    icons,
    assets,
  };
}
