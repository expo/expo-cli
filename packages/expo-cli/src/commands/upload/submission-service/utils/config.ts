import { ExpoConfig, getConfig } from '@expo/config';

interface AppConfig {
  owner?: string;
  slug: string;
}

function getAppConfig(projecDir: string): AppConfig {
  const exp = getExpoConfig(projecDir);
  if (!exp.slug) {
    throw new Error('expo.slug is not defined in app.json');
  }
  return {
    owner: exp.owner,
    slug: exp.slug,
  };
}

let exp: ExpoConfig;
function getExpoConfig(projectRoot: string): ExpoConfig {
  if (!exp) {
    const { exp: _exp } = getConfig(projectRoot, { skipSDKVersionRequirement: true });
    exp = _exp;
  }
  return exp;
}

export { getAppConfig, getExpoConfig };
