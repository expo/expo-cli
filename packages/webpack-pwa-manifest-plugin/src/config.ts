import { ExpoConfig } from '@expo/config';
import isColor from 'is-color';

import { PresetError } from './Errors';
import validatePresets from './validators/Presets';
import { ManifestOptions } from './WebpackPWAManifestPlugin.types';

function isObject(item: any): boolean {
  return typeof item === 'object' && !Array.isArray(item) && item !== null;
}

export function createPWAManifestFromExpoConfig(appJson: ExpoConfig) {
  if (!isObject(appJson)) {
    throw new Error('app.json must be an object');
  }

  const { web = {} } = appJson.expo || appJson || {};

  return {
    // PWA
    background_color: web.backgroundColor,
    description: web.description,
    dir: web.dir,
    display: web.display,
    lang: web.lang,
    name: web.name,
    orientation: web.orientation,
    prefer_related_applications: web.preferRelatedApplications,
    related_applications: web.relatedApplications,
    scope: web.scope,
    short_name: web.shortName,
    start_url: web.startUrl,
    theme_color: web.themeColor,
    crossorigin: web.crossorigin,
    startupImages: web.startupImages,
    icons: web.icons,
  };
}

export function validateManifest(manifest: ManifestOptions): void {
  if (!manifest) return;

  validatePresets(manifest, 'dir', 'display', 'orientation', 'crossorigin');

  for (const property of ['background_color', 'theme_color']) {
    // @ts-ignore
    const color = manifest[property];
    if (color && !isColor(color)) {
      throw new PresetError(property, color);
    }
  }
}
