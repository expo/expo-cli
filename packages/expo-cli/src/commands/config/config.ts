import { getConfig, ProjectConfig } from '@expo/config';
import { compileModsAsync } from '@expo/config-plugins/build/plugins/mod-compiler';
import { getPrebuildConfig } from '@expo/prebuild-config';
import type { Command } from 'commander';

import CommandError from '../../CommandError';
import Log from '../../log';
import { logConfig } from '../eject/configureProjectAsync';
import { profileMethod } from '../utils/profileMethod';

type Options = {
  type?: string;
  full: boolean;
};

async function actionAsync(projectRoot: string, options: Options) {
  let config: ProjectConfig;

  if (options.type === 'prebuild') {
    config = profileMethod(getPrebuildConfig)(projectRoot, {
      platforms: ['ios', 'android'],
    });
  } else if (options.type === 'introspect') {
    config = profileMethod(getPrebuildConfig)(projectRoot, {
      platforms: ['ios', 'android'],
    });

    await compileModsAsync(config.exp, {
      projectRoot,
      introspect: true,
      platforms: ['ios', 'android'],
      assertMissingModProviders: false,
    });
    // @ts-ignore
    delete config.modRequest;
    // @ts-ignore
    delete config.modResults;
  } else if (options.type === 'public') {
    config = profileMethod(getConfig)(projectRoot, {
      skipSDKVersionRequirement: true,
      isPublicConfig: true,
    });
  } else if (options.type) {
    throw new CommandError(
      `Invalid option: --type ${options.type}. Valid options are: public, prebuild`
    );
  } else {
    config = profileMethod(getConfig)(projectRoot, {
      skipSDKVersionRequirement: true,
    });
  }

  Log.log();
  logConfig(options.full ? config : config.exp);
  Log.log();
}

export default function (program: Command) {
  program
    .command('config [path]')
    .description('Show the project config')
    .helpGroup('info')
    .option('-t, --type <type>', 'Type of config to show. Options: public, prebuild, introspect')
    .option('--full', 'Include all project config data')
    .asyncActionProjectDir(profileMethod(actionAsync));
}
