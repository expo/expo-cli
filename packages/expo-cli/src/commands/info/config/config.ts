import type { Command } from 'commander';

import { applyAsyncActionProjectDir } from '../../utils/applyAsyncAction';

export default function (program: Command) {
  applyAsyncActionProjectDir(
    program
      .command('config [path]')
      .description('Show the project config')
      .helpGroup('info')
      .option('-t, --type <public|prebuild|introspect>', 'Type of config to show.')
      .option('--full', 'Include all project config data')
      .option('--json', 'Output in JSON format'),
    () => import('./configAsync')
  );
}
