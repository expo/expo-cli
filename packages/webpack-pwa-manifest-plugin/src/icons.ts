import { ImageFormat, ResizeMode, isAvailableAsync, sharpAsync } from '@expo/image-utils';
import fs from 'fs-extra';
import mime from 'mime';
import fetch from 'node-fetch';
import path from 'path';
import stream from 'stream';
import temporary from 'tempy';
import util from 'util';
import chalk from 'chalk';

import crypto from 'crypto';
import { IconError } from './Errors';
import { AnySize, ImageSize, joinURI, toArray, toSize } from './utils';
import { fromStartupImage } from './validators/Apple';
import { Icon, ManifestIcon, ManifestOptions } from './WebpackPWAManifestPlugin.types';
import { resize as jimpResize } from './ImageComposite';

const supportedMimeTypes = ['image/png', 'image/jpeg', 'image/webp'];

function sanitizeIcon(iconSnippet: Icon & { src?: string; size: any; sizes: any }): Icon {
  if (!iconSnippet.src) {
    throw new IconError('Unknown icon source.');
  }
  const sizes = toArray(iconSnippet.size || iconSnippet.sizes);
  if (!sizes) {
    throw new IconError('Unknown icon sizes.');
  }
  return {
    src: iconSnippet.src,
    resizeMode: iconSnippet.resizeMode,
    sizes,
    media: iconSnippet.media,
    destination: iconSnippet.destination,
    ios: iconSnippet.ios,
    color: iconSnippet.color,
  };
}

interface WebpackAsset {
  output: string;
  url: string;
  source: any;
  size: any;
  ios:
    | boolean
    | {
        valid: any;
        media: any;
        size: string;
        href: string;
      };
  resizeMode: any;
  color: any;
}

async function getBufferWithMimeAsync(
  { src, resizeMode, color }: Icon,
  mimeType: string,
  { width, height }: ImageSize
): Promise<Buffer> {
  let imagePath;
  if (!supportedMimeTypes.includes(mimeType)) {
    imagePath = src;
  } else {
    let localSrc = src;

    // In case the icon is a remote URL we need to download it first
    if (src.startsWith('http')) {
      localSrc = await downloadImage(src);
    }

    const imageData = await resize(localSrc, mimeType, width, height, resizeMode, color!);
    if (imageData instanceof Buffer) {
      return imageData;
    } else {
      imagePath = imageData;
    }
  }
  try {
    return await fs.readFile(imagePath);
  } catch (err) {
    throw new IconError(`It was not possible to read '${src}'.`);
  }
}

async function downloadImage(url: string): Promise<string> {
  const outputPath = temporary.directory();
  const localPath = path.join(outputPath, path.basename(url));

  const response = await fetch(url);
  if (!response.ok) {
    throw new IconError(`It was not possible to download splash screen from '${url}'`);
  }

  // Download to local file
  const streamPipeline = util.promisify(stream.pipeline);
  await streamPipeline(response.body, fs.createWriteStream(localPath));

  return localPath;
}

async function ensureCacheDirectory(projectRoot: string, cacheKey: string): Promise<string> {
  const cacheFolder = path.join(projectRoot, '.expo/web/cache/production/images', cacheKey);
  await fs.ensureDir(cacheFolder);
  return cacheFolder;
}

async function getImageFromCacheAsync(fileName: string, cacheKey: string): Promise<null | Buffer> {
  try {
    return await fs.readFile(path.resolve(cacheKeys[cacheKey], fileName));
  } catch (_) {
    return null;
  }
}

async function cacheImageAsync(fileName: string, buffer: Buffer, cacheKey: string): Promise<void> {
  try {
    await fs.writeFile(path.resolve(cacheKeys[cacheKey], fileName), buffer);
  } catch ({ message }) {
    console.warn(`error caching image: "${fileName}". ${message}`);
  }
}

let hasWarned: boolean = false;

async function processImageAsync(
  size: AnySize,
  icon: Icon,
  publicPath: string,
  cacheKey: string
): Promise<{ manifestIcon: ManifestIcon; webpackAsset: WebpackAsset }> {
  const { width, height } = toSize(size);
  if (width <= 0 || height <= 0) {
    throw Error(`Failed to process image with invalid size: { width: ${width}, height: ${height}}`);
  }
  const mimeType = mime.getType(icon.src);
  if (!mimeType) {
    throw new Error(`Invalid mimeType for image with source: ${icon.src}`);
  }

  const dimensions = `${width}x${height}`;
  const fileName = `icon_${dimensions}.${mime.getExtension(mimeType)}`;

  let imageBuffer: Buffer | null = await getImageFromCacheAsync(fileName, cacheKey);
  if (!imageBuffer) {
    // Putting the warning here will prevent the warning from showing if all images were reused from the cache
    if (!hasWarned && !(await isAvailableAsync())) {
      hasWarned = true;
      // TODO: Bacon: Fallback to nodejs image resizing as native doesn't work in the host environment.
      console.log('ff', cacheKey, fileName, dimensions);
      console.log();
      console.log(
        chalk.bgYellow.black(
          `PWA Images: Using node to generate images. This is much slower than using native packages.`
        )
      );
      console.log(
        chalk.yellow(
          `- Optionally you can stop the process and try again after successfully running \`npm install -g sharp-cli\`.\n- If you are using \`expo-cli\` to build your project then you could use the \`--no-pwa\` flag to skip the PWA asset generation step entirely.`
        )
      );
    }
    imageBuffer = await getBufferWithMimeAsync(icon, mimeType, { width, height });
    await cacheImageAsync(fileName, imageBuffer, cacheKey);
  }

  const iconOutputDir = icon.destination ? joinURI(icon.destination, fileName) : fileName;
  const iconPublicUrl = joinURI(publicPath, iconOutputDir);

  return {
    manifestIcon: {
      src: iconPublicUrl,
      sizes: dimensions,
      type: mimeType,
    },
    webpackAsset: {
      output: iconOutputDir,
      url: iconPublicUrl,
      source: imageBuffer,
      size: imageBuffer.length,
      ios: icon.ios
        ? { valid: icon.ios, media: icon.media, size: dimensions, href: iconPublicUrl }
        : false,
      resizeMode: icon.resizeMode,
      color: icon.color,
    },
  };
}

function ensureValidMimeType(mimeType: string): ImageFormat {
  if (['input', 'jpeg', 'jpg', 'png', 'raw', 'tiff', 'webp'].includes(mimeType)) {
    return mimeType as ImageFormat;
  }
  return 'png';
}

async function resize(
  inputPath: string,
  mimeType: string,
  width: number,
  height: number,
  fit: ResizeMode = 'contain',
  background: string
): Promise<string | Buffer> {
  if (!(await isAvailableAsync())) {
    return await jimpResize(inputPath, mimeType, width, height, fit, background);
  }

  const format = ensureValidMimeType(mimeType.split('/')[1]);
  const outputPath = temporary.directory();

  try {
    await sharpAsync(
      {
        input: inputPath,
        output: outputPath,
        format,
      },
      [
        {
          operation: 'flatten',
          background,
        },
        {
          operation: 'resize',
          width,
          height,
          fit,
          background,
        },
      ]
    );

    return path.join(outputPath, path.basename(inputPath));
  } catch ({ message }) {
    throw new IconError(`It was not possible to generate splash screen '${inputPath}'. ${message}`);
  }
}

export function retrieveIcons(manifest: ManifestOptions): [Icon[], ManifestOptions] {
  // Remove these items so they aren't written to disk.
  const { startupImages, icons, ...config } = manifest;
  const parsedStartupImages = toArray(startupImages);

  let parsedIcons = toArray(icons);

  if (parsedStartupImages.length) {
    // TODO: Bacon: use all of the startup images
    const startupImage = parsedStartupImages[0];
    parsedIcons = [...parsedIcons, ...fromStartupImage(startupImage)];
  }

  const response: Icon[] = parsedIcons.map(icon => sanitizeIcon(icon));
  return [response, config];
}

// Calculate SHA256 Checksum value of a file based on its contents
function calculateHash(filePath: string): string {
  const contents = filePath.startsWith('http') ? filePath : fs.readFileSync(filePath);

  return crypto
    .createHash('sha256')
    .update(contents)
    .digest('hex');
}

// Create a hash key for caching the images between builds
function createCacheKey(icon: Icon): string {
  const hash = calculateHash(icon.src);
  return [hash, icon.resizeMode, icon.color].filter(Boolean).join('-');
}

const cacheKeys: { [key: string]: string } = {};

export async function parseIconsAsync(
  projectRoot: string,
  inputIcons: Icon[],
  publicPath: string
): Promise<{ icons?: ManifestIcon[]; assets?: WebpackAsset[] }> {
  if (!inputIcons.length) {
    return {};
  }

  const icons: ManifestIcon[] = [];
  const assets: WebpackAsset[] = [];

  let promises: Promise<any>[] = [];

  for (const icon of inputIcons) {
    const cacheKey = createCacheKey(icon);
    if (!(cacheKey in cacheKeys)) {
      cacheKeys[cacheKey] = await ensureCacheDirectory(projectRoot, cacheKey);
    }

    const { sizes } = icon;
    promises = [
      ...promises,
      ...sizes!.map(async size => {
        const { manifestIcon, webpackAsset } = await processImageAsync(
          size,
          icon,
          publicPath,
          cacheKey
        );
        icons.push(manifestIcon);
        assets.push(webpackAsset);
      }),
    ];
  }
  await Promise.all(promises);

  await clearUnusedCachesAsync(projectRoot);

  return {
    icons: sortByAttribute(icons, 'sizes'),
    // startupImages: icons.filter(({ isStartupImage }) => isStartupImage),
    assets: sortByAttribute(assets, 'output'),
  };
}

function sortByAttribute(arr: any[], key: string): any[] {
  return arr.filter(Boolean).sort((valueA, valueB) => {
    if (valueA[key] < valueB[key]) return -1;
    else if (valueA[key] > valueB[key]) return 1;
    return 0;
  });
}

async function clearUnusedCachesAsync(projectRoot: string): Promise<void> {
  // Clean up any old caches
  const cacheFolder = path.join(projectRoot, '.expo/web/cache/production/images');
  const currentCaches = fs.readdirSync(cacheFolder);

  if (!Array.isArray(currentCaches)) {
    console.warn('Failed to read the icon cache');
    return;
  }
  const deleteCachePromises: Promise<void>[] = [];
  for (const cache of currentCaches) {
    // skip hidden folders
    if (cache.startsWith('.')) {
      continue;
    }

    // delete
    if (!(cache in cacheKeys)) {
      deleteCachePromises.push(fs.remove(path.join(cacheFolder, cache)));
    }
  }

  await Promise.all(deleteCachePromises);
}
