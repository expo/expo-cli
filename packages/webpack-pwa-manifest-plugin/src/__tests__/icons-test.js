import { getConfigForPWA } from '@expo/config';
import { resolve } from 'path';
import { parseIconsAsync, retrieveIcons } from '../icons';

import { createPWAManifestFromExpoConfig } from '../config';

it(`matches`, async () => {
  // Fill all config values with PWA defaults
  const config = getConfigForPWA(__dirname, (...props) => resolve(__dirname, ...props), {
    templateIcon: require.resolve('./icon.png'),
  });

  const manifest = createPWAManifestFromExpoConfig(config);

  const [icons] = retrieveIcons(manifest);

  const publicPath = '/DEMO_PATH';
  const { icons: parsedIcons, assets } = await parseIconsAsync(__dirname, icons, publicPath);

  for (const icon of parsedIcons) {
    const { sizes } = icon;
    // Ensure format: 100x200
    expect(sizes.split('x').length).toBe(2);
    // Ensure format: icon_1334x750.png
    expect(icon.src.startsWith(publicPath)).toBe(true);
    expect(icon.src.includes(sizes)).toBe(true);
    expect(icon.src.includes('png')).toBe(true);
    expect(icon.type).toBe('image/png');
  }

  const parsedAssets = assets.map(({ source, ...asset }) => ({ ...asset, source: '<removed>' }));

  expect(parsedAssets).toMatchSnapshot();

  // Ensure the splash props are being used from the app.json splash object
  const startupImages = parsedAssets.filter(({ ios }) => ios && ios.valid === 'startup');

  for (const image of startupImages) {
    expect(image.color).toBe(config.splash.backgroundColor);
    expect(image.resizeMode).toBe(config.splash.resizeMode);
  }
});
