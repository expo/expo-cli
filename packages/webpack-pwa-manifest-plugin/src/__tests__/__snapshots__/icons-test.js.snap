// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches 1`] = `
Array [
  Object {
    "color": undefined,
    "ios": Object {
      "href": "/DEMO_PATH/apple/icons/icon_180x180.png",
      "media": undefined,
      "size": "180x180",
      "valid": true,
    },
    "output": "apple/icons/icon_180x180.png",
    "resizeMode": undefined,
    "size": 4670,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/icons/icon_180x180.png",
  },
  Object {
    "color": undefined,
    "ios": false,
    "output": "apple/icons/icon_192x192.png",
    "resizeMode": undefined,
    "size": 4926,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/icons/icon_192x192.png",
  },
  Object {
    "color": undefined,
    "ios": false,
    "output": "apple/icons/icon_512x512.png",
    "resizeMode": undefined,
    "size": 8995,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/icons/icon_512x512.png",
  },
  Object {
    "color": "#ffffff",
    "ios": Object {
      "href": "/DEMO_PATH/apple/splash/icon_1125x2436.png",
      "media": "screen and (device-width: 375px) and (device-height: 812px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)",
      "size": "1125x2436",
      "valid": "startup",
    },
    "output": "apple/splash/icon_1125x2436.png",
    "resizeMode": "contain",
    "size": 64781,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/splash/icon_1125x2436.png",
  },
  Object {
    "color": "#ffffff",
    "ios": Object {
      "href": "/DEMO_PATH/apple/splash/icon_1136x640.png",
      "media": "screen and (device-width: 320px) and (device-height: 568px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)",
      "size": "1136x640",
      "valid": "startup",
    },
    "output": "apple/splash/icon_1136x640.png",
    "resizeMode": "contain",
    "size": 24359,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/splash/icon_1136x640.png",
  },
  Object {
    "color": "#ffffff",
    "ios": Object {
      "href": "/DEMO_PATH/apple/splash/icon_1242x2208.png",
      "media": "screen and (device-width: 414px) and (device-height: 736px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)",
      "size": "1242x2208",
      "valid": "startup",
    },
    "output": "apple/splash/icon_1242x2208.png",
    "resizeMode": "contain",
    "size": 75094,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/splash/icon_1242x2208.png",
  },
  Object {
    "color": "#ffffff",
    "ios": Object {
      "href": "/DEMO_PATH/apple/splash/icon_1242x2688.png",
      "media": "screen and (device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)",
      "size": "1242x2688",
      "valid": "startup",
    },
    "output": "apple/splash/icon_1242x2688.png",
    "resizeMode": "contain",
    "size": 77614,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/splash/icon_1242x2688.png",
  },
  Object {
    "color": "#ffffff",
    "ios": Object {
      "href": "/DEMO_PATH/apple/splash/icon_1334x750.png",
      "media": "screen and (device-width: 375px) and (device-height: 667px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)",
      "size": "1334x750",
      "valid": "startup",
    },
    "output": "apple/splash/icon_1334x750.png",
    "resizeMode": "contain",
    "size": 32416,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/splash/icon_1334x750.png",
  },
  Object {
    "color": "#ffffff",
    "ios": Object {
      "href": "/DEMO_PATH/apple/splash/icon_1792x828.png",
      "media": "screen and (device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)",
      "size": "1792x828",
      "valid": "startup",
    },
    "output": "apple/splash/icon_1792x828.png",
    "resizeMode": "contain",
    "size": 39849,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/splash/icon_1792x828.png",
  },
  Object {
    "color": "#ffffff",
    "ios": Object {
      "href": "/DEMO_PATH/apple/splash/icon_2208x1242.png",
      "media": "screen and (device-width: 414px) and (device-height: 736px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)",
      "size": "2208x1242",
      "valid": "startup",
    },
    "output": "apple/splash/icon_2208x1242.png",
    "resizeMode": "contain",
    "size": 80937,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/splash/icon_2208x1242.png",
  },
  Object {
    "color": "#ffffff",
    "ios": Object {
      "href": "/DEMO_PATH/apple/splash/icon_2436x1125.png",
      "media": "screen and (device-width: 375px) and (device-height: 812px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)",
      "size": "2436x1125",
      "valid": "startup",
    },
    "output": "apple/splash/icon_2436x1125.png",
    "resizeMode": "contain",
    "size": 70929,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/splash/icon_2436x1125.png",
  },
  Object {
    "color": "#ffffff",
    "ios": Object {
      "href": "/DEMO_PATH/apple/splash/icon_2688x1242.png",
      "media": "screen and (device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)",
      "size": "2688x1242",
      "valid": "startup",
    },
    "output": "apple/splash/icon_2688x1242.png",
    "resizeMode": "contain",
    "size": 86346,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/splash/icon_2688x1242.png",
  },
  Object {
    "color": "#ffffff",
    "ios": Object {
      "href": "/DEMO_PATH/apple/splash/icon_640x1136.png",
      "media": "screen and (device-width: 320px) and (device-height: 568px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)",
      "size": "640x1136",
      "valid": "startup",
    },
    "output": "apple/splash/icon_640x1136.png",
    "resizeMode": "contain",
    "size": 22612,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/splash/icon_640x1136.png",
  },
  Object {
    "color": "#ffffff",
    "ios": Object {
      "href": "/DEMO_PATH/apple/splash/icon_750x1334.png",
      "media": "screen and (device-width: 375px) and (device-height: 667px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)",
      "size": "750x1334",
      "valid": "startup",
    },
    "output": "apple/splash/icon_750x1334.png",
    "resizeMode": "contain",
    "size": 30660,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/splash/icon_750x1334.png",
  },
  Object {
    "color": "#ffffff",
    "ios": Object {
      "href": "/DEMO_PATH/apple/splash/icon_828x1792.png",
      "media": "screen and (device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)",
      "size": "828x1792",
      "valid": "startup",
    },
    "output": "apple/splash/icon_828x1792.png",
    "resizeMode": "contain",
    "size": 37078,
    "source": "<removed>",
    "url": "/DEMO_PATH/apple/splash/icon_828x1792.png",
  },
]
`;
